src/sampling.jl

```diff
@@ -0,0 +1,177 @@
+using StatsBase
+"""
+	Contains all function for sampling expressions and from expressions
+"""
+
+"""
+    rand(::Type{RuleNode}, grammar::Grammar, typ::Symbol, max_depth::Int=10)
+
+Generates a random RuleNode of return type typ and maximum depth max_depth.
+"""
+function Base.rand(::Type{RuleNode}, grammar::Grammar, typ::Symbol, max_depth::Int=10, 
+    bin::Union{NodeRecycler,Nothing}=nothing)
+    rules = grammar[typ]
+    
+    if max_depth <= 1
+        terminals = filter(r->isterminal(grammar,r), rules)
+        rule_index = !isempty(terminals) ? StatsBase.sample(terminals) : StatsBase.sample(rules)
+    else    
+        rule_index = StatsBase.sample(rules)
+    end
+
+    rulenode = iseval(grammar, rule_index) ?
+        RuleNode(bin, rule_index, Core.eval(grammar, rule_index)) :
+        RuleNode(bin, rule_index)
+
+    if !grammar.isterminal[rule_index]
+        for ch in child_types(grammar, rule_index)
+            push!(rulenode.children, rand(RuleNode, grammar, ch, max_depth-1, bin))
+        end
+    end
+    return rulenode
+end
+"""
+    rand(::Type{RuleNode}, grammar::Grammar, typ::Symbol, dmap::AbstractVector{Int}, max_depth::Int=10)
+
+Generates a random RuleNode of return type typ and maximum depth max_depth guided by a minimum depth map dmap.
+"""
+function Base.rand(::Type{RuleNode}, grammar::Grammar, typ::Symbol, dmap::AbstractVector{Int}, 
+    max_depth::Int=10, bin::Union{NodeRecycler,Nothing}=nothing)
+    rules = grammar[typ]
```

Comment:
add constraint propagation as in here: https://github.com/Herb-AI/HerbSearch.jl/blob/debbc00ebc42782595539d9a105f74ccea25a418/src/csg_priority_enumerator.jl#L27

Reply:
See comment above